### PR TITLE
Fix Parallel.For non-deterministic slowdown in 32-bit processes

### PR DESCRIPTION
**Pull request by stephentoub**
Parallel.For uses the same code paths to handle its int and long-based overloads, using longs internally for both.  Interlocked operations are used to move within the ranges created to partition the input space, and those interlocked operations happen on longs.  In 32-bit processes, these long values may end up aligned on 4-byte but not 8-byte boundaries, in which case they can end up spanning two cache lines, which makes Interlocked.Add very slow, and can cause very measurable slow downs when using Parallel.For in a 32-bit process based on where the range objects are allocated and moved by the GC.

This commit fixes this by checking whether the current process is 32-bit and whether each partitioned range can fit in an int.  If both are true, we use Interlocked.Add(int, ...) instead of Interlocked.Add(long, ...), only falling back to using Interlocked.Add(long, ...) in a 32-bit process if you actually have enough input to warrant it (e.g. ranges of at least 2B items), which is very rare.

With this repro:
```C#
[Fact]
public static void Repro()
{
    var s = new Stopwatch();
    for (int t = 0; t < 10; t++)
    {
        var arr = new int[1024 * 1024];
        s.Restart();
        Parallel.For(0, 1024 * arr.Length, i => arr[i % arr.Length] = i);
        Console.WriteLine(s.ElapsedMilliseconds);
    }
}
```
Running it 10 times in a release build on 64-bit before and after the change:
```
Before
  1284
  1159
  1254
  1224
  1230
  1180
  1184
  1218
  1253
  1175

After
  1299
  1205
  1213
  1144
  1210
  1245
  1234
  1171
  1174
  1207
```
so basically no change.  But running it 10 times in a release build on 32-bit before and after the change:
```
Before
  1835
  2075
  24780
  2145
  2055
  2173
  1903
  1917
  25289
  2137

After
  1757
  2006
  1589
  1822
  1682
  1818
  1755
  1684
  1728
  1722
```
Note that random large jumps in execution time on several of the runs before and none of the runs after.

Fixes https://github.com/dotnet/corefx/issues/19276
cc: @kouvel, @AlexGhiondea, @alexperovich, @AlfredoMS

### CODE REVIEW

File: src/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/ParallelRangeManager.cs [hunk 1]
```diff
@@ -37,6 +41,7 @@ internal struct IndexRange
     /// <summary>
     /// The RangeWorker struct wraps the state needed by a task that services the parallel loop
     /// </summary>
+    [StructLayout(LayoutKind.Auto)]
```
Comment (danmosemsft) [inline src/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/ParallelRangeManager.cs]: is'nt this default?

Comment (stephentoub) [inline src/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/ParallelRangeManager.cs]: Not for structs. The default for a struct is Sequential.

Comment (danmosemsft) [inline src/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/ParallelRangeManager.cs]: what does marking it Auto achieve here? Is it just something you do by default when it's not used for interop?

Comment (stephentoub) [inline src/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/ParallelRangeManager.cs]: In this case, the struct has the following fields, in this defined order:
- reference
- int
- long
- long
- long
- bool

In a 64-bit process, with sequential layout, the int field will need to be padded so that the longs can be properly aligned, and the bool will end up being expanded to be 8 bytes as well, so the struct will be 48 bytes.  As auto, the runtime can reorder the fields to avoid the extra padding, e.g. putting the int at the end with the bool, and it'll end up being only 40 bytes.

Comment (stephentoub) [inline src/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/ParallelRangeManager.cs]: (But generally, yes, it's something I typically do by default, when I remember and when it's not inappropriate for some reason, so that I don't need to worry about the size and order of the fields as much and can let the runtime take care of an optimal ordering for me.)

Comment (danmosemsft) [inline src/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/ParallelRangeManager.cs]: OK good to learn thx.

**Comment by danmosemsft** [comment]
I'm disabling that flaky test in https://github.com/dotnet/corefx/pull/19256.

**Comment by danmosemsft** [comment]
@dotnet-bot test Innerloop OSX10.12 Debug x64 Build and Test  please (#19260)

File: src/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/ParallelRangeManager.cs [hunk 2]
```diff
@@ -99,9 +108,24 @@ internal bool FindNewWork(out long nFromInclusiveLocal, out long nToExclusiveLoc
...
-                                                    _nIncrementValue) - _nIncrementValue;
+                    long nMyOffset;
+                    if (IntPtr.Size == 4 && _use32BitCurrentIndex)
+                    {
+                        // In 32-bit processes, we prefer to use 32-bit interlocked operations, to avoid the possibility of doing
+                        // a 64-bit interlocked when the target value crosses a cache line, as that can be super expensive.
+                        // We use the first 32 bits of the Int64 index in such cases.
+                        unsafe
+                        {
+                            fixed (long* indexPtr = &_indexRanges[_nCurrentIndexRange]._nSharedCurrentIndexOffset.Value)
```
Comment (davidfowl) [inline src/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/ParallelRangeManager.cs]: Could you use ref locals here instead of pinning (of course you're just replacing unsafe with the Unsafe library to do the cast).

```C#
ref int index = ref Unsafe.As<long, int>(ref currentRange._nSharedCurrentIndexOffset.Value);
nMyOffset = Interlocked.Add(ref index, (int)_nIncrementValue) - _nIncrementValue;
```

Comment (stephentoub) [inline src/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/ParallelRangeManager.cs]: Probably, but that would also require adding a reference to the Unsafe library, which I was hesitant to do because this assembly is in netcore's netstandard implementation and Unsafe currently targets netstandard.  I'm not particularly concerned with the fixed here (it's very short), and we can revise it if we like once all of the Unsafe/Span stuff is propagates more throughout corefx.
